keps/672-execution-policy/README.md

```diff
@@ -0,0 +1,383 @@
+                    - mountPath: /workspace/pre-trained-model
+                      name: model-initializer
+              volumes:
+                - name: dataset-initializer
+                  persistentVolumeClaim:
+                    claimName: dataset-initializer
+                - name: model-initializer
+                  persistentVolumeClaim:
+                    claimName: model-initializer
+    - name: Node
+      template:
+        spec:
+          parallelism: 3
+          completions: 3
+          template:
+            spec:
+              containers:
+                - name: trainer
+                  image: docker.io/kubeflow/deepspeed-trainer
+                  resources:
+                    limits:
+                      nvidia.com/gpu: 5
+                  volumeMounts:
+                    - mountPath: /workspace/dataset
+                      name: dataset-initializer
+                    - mountPath: /workspace/pre-trained-model
+                      name: model-initializer
+              volumes:
+                - name: dataset-initializer
+                  persistentVolumeClaim:
+                    claimName: dataset-initializer
+                - name: model-initializer
+                  persistentVolumeClaim:
+                    claimName: model-initializer
+```
+
+#### Story 3
+
+TODO: Add HPC use-case with Job sequence
+
+### Risks and Mitigations
+
+This API will not allow to describe DAGs to avoid workflow manager features in JobSet.
+The goal is to only focus on Job sequence to cover model training/HPC use-cases.
+
+## Design Details
+
+### API Details
+
+```golang
+type JobSetSpec struct {
+	ExecutionPolicy *ExecutionPolicy `json:"executionPolicy,omitempty"`
+}
+
+type ExecutionPolicyOption string
+
+const (
+	// This is the default settings.
+	// AnyOrder means that Jobs will be started in any order.
+	AnyOrder ExecutionPolicyOption = "AnyOrder"
+
+	// InOrder starts the ReplicatedJobs in order that they are listed. Jobs within a ReplicatedJob
+	// will still start in any order.
+	InOrder ExecutionPolicyOption = "InOrder"
+)
+
+type ExecutionPolicy struct {
+	// Order in which Jobs will be created.
+	ExecutionPolicyOrder ExecutionPolicyOption `json:"executionPolicyOrder"`
+
+	// After all ReplicatedJobs reach this status, the JobSet will create the next ReplicatedJobs.
+	ExecutionPolicyRule []ExecutionPolicyRule `json:"rules"`
+}
+
+// ExecutionPolicyRule represents the execution policy rule for Job sequence.
+type ExecutionPolicyRule struct {
+
+	// Names of the replicated Jobs that applied the status.
+	TargetReplicatedJobs []string `json:"targetReplicatedJobs"`
+
+	// Status in which the next replicated Jobs will be created.
+	ReplicatedJobsStatus ReplicatedJobsStatusOption `json:"replicatedJobsStatus"`
+}
+
+type ReplicatedJobsStatusOption string
+
+// TODO: What statuses do we want to support in the first version ?
+const (
+	ReadyStatus ReplicatedJobsStatusOption = "Ready"
```

Comment:
If we use an execution policy which specifies prior replicatedJobs must have a "Ready" status before subsequent ones can execute, isn't this effectively the same thing as using a StartupPolicy?

If so, perhaps we could eventually deprecate StartupPolicy in favor of this more flexible ExecutionPolicy over a couple releases

Reply:
>If we use an execution policy which specifies prior replicatedJobs must have a "Ready" status before subsequent ones can execute, isn't this effectively the same thing as using a StartupPolicy?

That is correct. Eventually, we just run the StartupPolicy functionality here.

>If so, perhaps we could eventually deprecate StartupPolicy in favor of this more flexible ExecutionPolicy over a couple releases

I am fine with that, but want to hear @kannon92 @ahg-g @vsoch @tenzen-y opinion on that.

Reply:
`StartupPolicy` seems more intuitive (from a user perspective) than an `ExecutionPolicy` -> Ready state, but I do think the latter is more flexible. I don't have a strong opinion either way, but if something is deprecated it should be well documented, etc.

Reply:
I think we need to phase out StartupPolicy over a few releases.

So I think this KEP should comment on compatibility between StartupPolicy and ExecutionPolicy.

If I specify StartupPolicy and execution Policy together, what would happen?  Should we enforce validation so that these are "either or".

Reply:
>If I specify StartupPolicy and execution Policy together, what would happen? Should we enforce validation so that these are "either or".

Yes, I think we should enforce "either or" on the first release then deprecate startup policy over the next few releases.

Reply:
Sure, @kannon92 @danielvegamyhre do we need separate section to explain it ?
Currently, I explained it in the validation section:
```
StartupPolicy should be equal to AnyOrder when ExecutionPolicy is used.
The StartupPolicy API will be deprecated over the next few JobSet releases, since
ExecutionPolicy can be used with Ready status of ReplicatedJobs.
```

Reply:
Yes, I mentioned it in one of the alternatives: https://github.com/kubernetes-sigs/jobset/blob/4a77dab7aabbc99fee259f61f6b4e4c7efb51470/keps/672-execution-policy/README.md#add-executionpolicyrule-parameter-into-the-startuppolicy-api.